scio-smb/src/main/scala/com/spotify/scio/smb/syntax/SortMergeBucketScioContextSyntax.scala

```diff
@@ -0,0 +1,144 @@
+   * `lhs` and (k, v2) is in `rhs`.
+   *
+   * Unlike a regular [[PairSCollectionFunctions.join()]], the key information (namely, how to
+   * extract a comparable `K` from `L` and `R`) is remotely encoded in a
+   * [[org.apache.beam.sdk.extensions.smb.BucketMetadata]] file in the same directory as the
+   * input records. This transform requires a filesystem lookup to ensure that the metadata for
+   * each source are compatible.
+   *
+   * @group join
+
+   * @param keyClass join key class. Must have a Coder in Beam's default
+   *                 [[org.apache.beam.sdk.coders.CoderRegistry]] as custom key coders are not
+   *                 supported yet.
+   * @param lhs
+   * @param rhs
+   */
+  def sortMergeJoin[K: Coder, L: Coder, R: Coder](
+    keyClass: Class[K],
+    lhs: SortedBucketIO.Read[L],
+    rhs: SortedBucketIO.Read[R]
+  ): SCollection[(K, (L, R))] =
+    sortMergeCoGroup(keyClass, lhs, rhs)
+      .flatMap {
+        case (k, (l, r)) =>
+          for {
+            i <- l
+            j <- r
+          } yield (k, (i, j))
+      }
+
+  /**
+   * For each key K in `a` or `b` return a resulting SCollection that contains a tuple with the
+   * list of values for that key in `a`, and `b`.
+   *
+   * See note on [[SortedBucketScioContext.sortMergeJoin() for information on how an SMB cogroup
+   * differs from a regular [[org.apache.beam.sdk.schemas.transforms.CoGroup]] operation.
+   *
+   * @group cogroup
+
+   * @param keyClass cogroup key class. Must have a Coder in Beam's default
+   *                 [[org.apache.beam.sdk.coders.CoderRegistry]] as custom key coders are not
+   *                 supported yet.
+   */
+  def sortMergeCoGroup[K: Coder, A: Coder, B: Coder](
+    keyClass: Class[K],
+    a: SortedBucketIO.Read[A],
+    b: SortedBucketIO.Read[B]
+  ): SCollection[(K, (Iterable[A], Iterable[B]))] = {
+    val t = SortedBucketIO.read(keyClass).of(a).and(b)
+    val (tupleTagA, tupleTagB) = (
+      a.getTupleTag,
+      b.getTupleTag
+    )
+    self
+      .wrap(self.applyInternal(t))
+      .map { kv =>
+        val cgbkResult = kv.getValue
+
+        (
+          kv.getKey,
+          (
+            cgbkResult.getAll(tupleTagA).asScala,
+            cgbkResult.getAll(tupleTagB).asScala
+          )
+        )
+      }
+  }
+
+  /**
+   * For each key K in `a` or `b` or `c`, return a resulting SCollection that contains a tuple
+   * with the list of values for that key in `a`, `b` and `c`.
+   *
+   * See note on [[SortedBucketScioContext.sortMergeJoin() for information on how an SMB cogroup
+   * differs from a regular [[org.apache.beam.sdk.schemas.transforms.CoGroup]] operation.
+   *
+   * @group cogroup
+
+   * @param keyClass cogroup key class. Must have a Coder in Beam's default
+   *                 [[org.apache.beam.sdk.coders.CoderRegistry]] as custom key coders are not
+   *                 supported yet.
+   */
+  def sortMergeCoGroup[K: Coder, A: Coder, B: Coder, C: Coder](
```

Comment:
Let's add a 4-way one to match `PairSCollectionFunctions`, and save `MultiSmb` for later?